import os
from PyQt5 import uic
from PyQt5.QtWidgets import QMainWindow, QFileDialog, QMessageBox, QMenu, QInputDialog, QTreeWidgetItem
from PyQt5.QtCore import Qt
from PyQt5.QtGui import QTextCursor
from src.managers.file_manager import FileManager
from src.managers.terminal_manager import TerminalManager
from src.managers.model_manager import ModelManager
from src.managers.agent_manager import AgentManager

# Load UI file
form_class = uic.loadUiType("./ui/editor.ui")[0]

"""ì—ë””í„° ë©”ì¸ ìœˆë„ìš° í´ë˜ìŠ¤"""
class EditorWindow(QMainWindow, form_class):
    def __init__(self):
        super().__init__()
        self.setupUi(self)
        
        # Manager ì´ˆê¸°í™”
        self.file_manager = FileManager()
        self.terminal_manager = TerminalManager()
        self.model_manager = ModelManager()
        self.agent_manager = AgentManager(self.model_manager)

        # ëª¨ë¸ ì½¤ë³´ë°•ìŠ¤ ì´ˆê¸°í™”
        self._populate_model_combobox()
        
        # global ìƒíƒœ ë³€ìˆ˜
        self.opened_folder_path = ""
        self.opened_file_path = ""
        self.opened_file_name = ""
        
        # Initialize UI
        self.filename_label.setText("Untitled")
        self.file_list.clear()
        self.file_list.setContextMenuPolicy(Qt.CustomContextMenu)

        # Hotkeys
        self.actionOpen.setShortcut("Ctrl+O")
        self.actionSave.setShortcut("Ctrl+S")
        self.actionExit.setShortcut("Ctrl+Q")
        
        # í„°ë¯¸ë„ ì´ˆê¸° ë©”ì‹œì§€
        self.terminal_output.setPlainText("í„°ë¯¸ë„ì´ ì¤€ë¹„ë˜ì—ˆìŠµë‹ˆë‹¤. í´ë”ë¥¼ ì—´ë©´ í•´ë‹¹ ë””ë ‰í† ë¦¬ì—ì„œ ì‹œì‘ë©ë‹ˆë‹¤.\n")
        
        # ì´ˆê¸° ìƒíƒœ: í´ë” ì—´ê¸° ë²„íŠ¼ í‘œì‹œ
        self.file_list_stack.setCurrentIndex(0)
        
        # Connect events
        self._connect_events()
        # í„°ë¯¸ë„ ì¶œë ¥ ìŠ¤íŠ¸ë¦¼ ì—°ê²°
        self.terminal_manager.output_received.connect(self.append_terminal_output)
    
    """ì´ë²¤íŠ¸ ì—°ê²°"""
    def _connect_events(self):
        self.actionOpen.triggered.connect(self.open_folder)
        self.actionSave.triggered.connect(self.save_file)
        self.actionExit.triggered.connect(self.close)
        self.file_list.itemClicked.connect(self.open_file)
        self.file_list.customContextMenuRequested.connect(self.show_context_menu)
        self.terminal_input.returnPressed.connect(self.execute_terminal_command)
        self.open_folder_button.clicked.connect(self.open_folder)
        self.agent_enterButton.clicked.connect(self.on_agent_generate)
        self.modelName.currentTextChanged.connect(self._on_model_selected)
    
    """í´ë” ì—´ê¸°"""
    def open_folder(self):
        folder_path = QFileDialog.getExistingDirectory(self, "Select Folder")
        
        if folder_path:
            self.opened_folder_path = folder_path
            print("Opened Folder:", self.opened_folder_path)
            
            self.file_list_stack.setCurrentIndex(1)
            
            self.file_manager.load_folder_tree(self.file_list, folder_path)
            
            # í„°ë¯¸ë„ ë””ë ‰í† ë¦¬ ì„¤ì •
            self.terminal_manager.start_terminal(folder_path)
            self.append_terminal_output(f"\nì‘ì—… ë””ë ‰í† ë¦¬: {folder_path}\n")
        else:
            print("No folder selected")
    
    """íŒŒì¼ ì—´ê¸°"""
    def open_file(self, item, column):
        if item.parent() is None and not hasattr(self, 'opened_folder_path'):
            return
        
        file_path = self.file_manager.get_file_path_from_item(item, self.opened_folder_path)
        
        if os.path.isfile(file_path):
            content, error = self.file_manager.read_file(file_path)
            
            if error:
                print(f"Error opening file: {error}")
                self.code_input.setPlainText("")
            else:
                self.code_input.setPlainText(content)
                self.opened_file_path = file_path
                self.opened_file_name = os.path.basename(file_path)
                self.filename_label.setText(self.opened_file_name)
                print(f"Opened file: {file_path}")
        else:
            print(f"Not a file: {file_path}")
    
    """íŒŒì¼ ì €ì¥"""
    def save_file(self):
        if not self.opened_file_path or not os.path.isfile(self.opened_file_path):
            print("No file to save")
            return
        
        content = self.code_input.toPlainText()
        success, error = self.file_manager.save_file(self.opened_file_path, content)
        
        if success:
            QMessageBox.information(self, "ì €ì¥ ì™„ë£Œ", f"íŒŒì¼ì´ ì €ì¥ë˜ì—ˆìŠµë‹ˆë‹¤:\n{self.opened_file_path}")
            print(f"File saved: {self.opened_file_path}")
        else:
            print(f"Error saving file: {error}")
    
    """íŒŒì¼ ë¦¬ìŠ¤íŠ¸ ìš°í´ë¦­ ì»¨í…ìŠ¤íŠ¸ ë©”ë‰´ í‘œì‹œ"""
    def show_context_menu(self, position):
        item = self.file_list.itemAt(position)
        if not item:
            return
        
        file_path = self.file_manager.get_file_path_from_item(item, self.opened_folder_path)
        is_directory = os.path.isdir(file_path)
        
        menu = QMenu(self)
        
        if is_directory:
            add_file_action = menu.addAction("íŒŒì¼ ì¶”ê°€")
            add_file_action.triggered.connect(lambda: self.add_file_to_folder(item, file_path))
            menu.addSeparator()
            add_folder_action = menu.addAction("í´ë” ì¶”ê°€")
            add_folder_action.triggered.connect(lambda: self.create_folder_in_folder(item, file_path))
            menu.addSeparator()
        
        delete_action = menu.addAction("ì‚­ì œ")
        delete_action.triggered.connect(lambda: self.delete_item(item, file_path, is_directory))
        
        menu.exec_(self.file_list.viewport().mapToGlobal(position))
    
    """í´ë”ì— ìƒˆ íŒŒì¼ ì¶”ê°€"""
    def add_file_to_folder(self, folder_item, folder_path):
        file_name, ok = QInputDialog.getText(self, "íŒŒì¼ ì¶”ê°€", "ìƒˆ íŒŒì¼ ì´ë¦„ì„ ì…ë ¥í•˜ì„¸ìš”:")
        
        if ok and file_name:
            new_file_path = os.path.join(folder_path, file_name)
            
            if os.path.exists(new_file_path):
                QMessageBox.warning(self, "ì˜¤ë¥˜", "ê°™ì€ ì´ë¦„ì˜ íŒŒì¼ì´ ì´ë¯¸ ì¡´ì¬í•©ë‹ˆë‹¤.")
                return
            
            success, error = self.file_manager.create_file(new_file_path)
            
            if success:
                new_item = QTreeWidgetItem(folder_item)
                new_item.setText(0, file_name)
                new_item.setData(0, 1, file_name)
                folder_item.setExpanded(True)
                
                QMessageBox.information(self, "ì„±ê³µ", f"íŒŒì¼ì´ ìƒì„±ë˜ì—ˆìŠµë‹ˆë‹¤:\n{new_file_path}")
                print(f"File created: {new_file_path}")
            else:
                QMessageBox.critical(self, "ì˜¤ë¥˜", f"íŒŒì¼ ìƒì„± ì‹¤íŒ¨:\n{error}")
    
    """í´ë”ì— ìƒˆ í´ë” ì¶”ê°€"""
    def create_folder_in_folder(self, folder_item, folder_path):
        folder_name, ok = QInputDialog.getText(self, "í´ë” ì¶”ê°€", "ìƒˆ í´ë” ì´ë¦„ì„ ì…ë ¥í•˜ì„¸ìš”:")
        
        if ok and folder_name:
            new_folder_path = os.path.join(folder_path, folder_name)
            
            if os.path.exists(new_folder_path):
                QMessageBox.warning(self, "ì˜¤ë¥˜", "ê°™ì€ ì´ë¦„ì˜ í´ë”ê°€ ì´ë¯¸ ì¡´ì¬í•©ë‹ˆë‹¤.")
                return
            
            success, error = self.file_manager.create_folder(new_folder_path)
            
            if success:
                new_item = QTreeWidgetItem(folder_item)
                new_item.setText(0, f"ğŸ“ {folder_name}")
                new_item.setData(0, 1, folder_name)
                folder_item.setExpanded(True)
                
                QMessageBox.information(self, "ì„±ê³µ", f"í´ë”ê°€ ìƒì„±ë˜ì—ˆìŠµë‹ˆë‹¤:\n{new_folder_path}")
                print(f"Folder created: {new_folder_path}")
            else:
                QMessageBox.critical(self, "ì˜¤ë¥˜", f"í´ë” ìƒì„± ì‹¤íŒ¨:\n{error}")

    """íŒŒì¼ ë˜ëŠ” í´ë” ì‚­ì œ"""
    def delete_item(self, item, item_path, is_directory):
        item_type = "í´ë”" if is_directory else "íŒŒì¼"
        reply = QMessageBox.question(
            self, 
            "ì‚­ì œ í™•ì¸", 
            f"ì •ë§ ì´ {item_type}ë¥¼ ì‚­ì œí•˜ì‹œê² ìŠµë‹ˆê¹Œ?\n{item_path}",
            QMessageBox.Yes | QMessageBox.No,
            QMessageBox.No
        )
        
        if reply == QMessageBox.Yes:
            success, error = self.file_manager.delete_item(item_path, is_directory)
            
            if success:
                parent = item.parent()
                if parent:
                    parent.removeChild(item)
                else:
                    index = self.file_list.indexOfTopLevelItem(item)
                    self.file_list.takeTopLevelItem(index)
                
                # í˜„ì¬ ì—´ë¦° íŒŒì¼ì´ ì‚­ì œëœ ê²½ìš° ì—ë””í„° ì´ˆê¸°í™”
                if self.opened_file_path == item_path:
                    self.code_input.setPlainText("")
                    self.filename_label.setText("Untitled")
                    self.opened_file_path = ""
                    self.opened_file_name = ""
                
                QMessageBox.information(self, "ì„±ê³µ", f"{item_type}ê°€ ì‚­ì œë˜ì—ˆìŠµë‹ˆë‹¤.")
                print(f"Deleted: {item_path}")
            else:
                QMessageBox.critical(self, "ì˜¤ë¥˜", f"{item_type} ì‚­ì œ ì‹¤íŒ¨:\n{error}")
    
    """í„°ë¯¸ë„ ëª…ë ¹ì–´ ì‹¤í–‰"""
    def execute_terminal_command(self):
        command = self.terminal_input.text().strip()
        if not command:
            return
        
        # í˜„ì¬ ë””ë ‰í† ë¦¬ í‘œì‹œ
        current_dir = self.terminal_manager.get_current_directory()
        prompt = f"PS {current_dir}> " if current_dir else "> "
        
        # ëª…ë ¹ì–´ í‘œì‹œ
        self.append_terminal_output(f"{prompt}{command}\n")
        self.terminal_input.clear()
        
        # cd
        if command.startswith("cd "):
            new_path = command[3:].strip().strip('"').strip("'")
            
            # ìƒëŒ€ ê²½ë¡œ
            if not os.path.isabs(new_path):
                new_path = os.path.join(current_dir, new_path)
            
            new_path = os.path.normpath(new_path)
            success, message = self.terminal_manager.change_directory(new_path)
            self.append_terminal_output(f"{message}\n")
        
        # clear
        elif command.lower() in ["clear", "cls"]:
            self.terminal_output.clear()
            self.append_terminal_output("í„°ë¯¸ë„ì´ ì´ˆê¸°í™”ë˜ì—ˆìŠµë‹ˆë‹¤.\n")
        
        # else
        else:
            # ì§€ì†í˜• ì…¸ì— ëª…ë ¹/ì…ë ¥ì„ ì „ë‹¬
            success, msg = self.terminal_manager.execute_command(command)
            if not success and msg:
                self.append_terminal_output(msg + "\n")
    
    """í„°ë¯¸ë„ ì¶œë ¥ì— í…ìŠ¤íŠ¸ ì¶”ê°€"""
    def append_terminal_output(self, text):
        self.terminal_output.moveCursor(QTextCursor.End)
        self.terminal_output.insertPlainText(text)
        self.terminal_output.moveCursor(QTextCursor.End)

    """ëª¨ë¸ ì½¤ë³´ë°•ìŠ¤ ì´ˆê¸°í™”"""
    def _populate_model_combobox(self):
        """ModelManagerì— ë“±ë¡ëœ ëª¨ë¸ë“¤ì„ `modelName` ì½¤ë³´ë°•ìŠ¤ì— ì±„ì›ë‹ˆë‹¤."""
        try:
            self.modelName.clear()
            # ëª¨ë¸ alias ëª©ë¡ì„ ì¶”ê°€
            for alias in sorted(self.model_manager.models.keys()):
                self.modelName.addItem(alias)

            # ê¸°ë³¸ ëª¨ë¸ ì„ íƒ
            default = getattr(self.model_manager, 'default_model_alias', None)
            if default:
                idx = self.modelName.findText(default)
                if idx >= 0:
                    self.modelName.setCurrentIndex(idx)
        except Exception as e:
            # UIì— ì˜¤ë¥˜ ì¶œë ¥
            self.append_terminal_output(f"ëª¨ë¸ ëª©ë¡ ë¡œë“œ ì‹¤íŒ¨: {e}\n")

    """ëª¨ë¸ ì½¤ë³´ë°•ìŠ¤ ì„ íƒ ë³€ê²½ í•¸ë“¤ëŸ¬"""
    def _on_model_selected(self, alias: str):
        """ì½¤ë³´ë°•ìŠ¤ ì„ íƒ ë³€ê²½ ì‹œ ModelManagerì˜ ê¸°ë³¸ ëª¨ë¸ì„ ë³€ê²½í•©ë‹ˆë‹¤."""
        if not alias:
            return
        try:
            self.model_manager.set_default_model(alias)
            self.append_terminal_output(f"ëª¨ë¸ ì„ íƒ: {alias}\n")
        except Exception as e:
            QMessageBox.warning(self, "ëª¨ë¸ ì„ íƒ ì˜¤ë¥˜", str(e))

    """ì—ì´ì „íŠ¸ ìƒì„± ì‹¤í–‰"""
    def on_agent_generate(self):
        # íŒŒì¼ì´ ì—´ë ¤ ìˆì–´ì•¼ í•¨
        if not self.opened_file_path:
            QMessageBox.warning(self, "ì—ì´ì „íŠ¸", "ë¨¼ì € íŒŒì¼ì„ ì—´ì–´ì£¼ì„¸ìš”.")
            return

        # ì‚¬ìš©ì í”„ë¡¬í”„íŠ¸ì™€ í˜„ì¬ ì½”ë“œ(ìˆ˜ì • ì „/ì¤‘)ë¥¼ ê°€ì ¸ì˜´
        user_prompt = self.agent_promptEdit.toPlainText().strip()
        if not user_prompt:
            QMessageBox.information(self, "ì—ì´ì „íŠ¸", "í”„ë¡¬í”„íŠ¸ë¥¼ ì…ë ¥í•´ì£¼ì„¸ìš”.")
            return
        
        self.agent_promptEdit.clear()

        current_code = self.code_input.toPlainText()

        # ì‹¤í–‰ ì¤‘ í‘œì‹œ
        self.agent_enterButton.setEnabled(False)
        self.agent_enterButton.setText("â€¦")
        self.agent_resultEdit.setPlainText("LLM í˜¸ì¶œ ì¤‘ì…ë‹ˆë‹¤â€¦")
        try:
            ok, raw, extracted_code, extracted_desc = self.agent_manager.run(user_prompt, self.opened_file_path, current_code)
        finally:
            self.agent_enterButton.setEnabled(True)
            self.agent_enterButton.setText("âœ¦")

        if not ok:
            self.agent_resultEdit.setHtml(self._format_as_html(raw))
            QMessageBox.critical(self, "ì—ì´ì „íŠ¸ ì˜¤ë¥˜", raw)
            return

        # ì„¤ëª… ë¸”ë¡ì´ ìˆìœ¼ë©´ ê·¸ê±¸ í‘œì‹œ, ì—†ìœ¼ë©´ ì „ì²´ ì‘ë‹µ í‘œì‹œ
        if extracted_desc:
            self.agent_resultEdit.setHtml(self._format_as_html(extracted_desc))
        else:
            self.agent_resultEdit.setHtml(self._format_as_html(raw))

        # í¬ë§·ì— ë§ëŠ” ì½”ë“œ ì¶”ì¶œ ì„±ê³µ ì‹œ ì½”ë“œ ì…ë ¥ì°½ì— ë°˜ì˜
        if extracted_code:
            self.code_input.setPlainText(extracted_code)
        else:
            # ì¶”ì¶œ ì‹¤íŒ¨ ì‹œ ì•ˆë‚´ ìœ ì§€
            QMessageBox.information(self, "ì•ˆë‚´", "ì‘ë‹µì—ì„œ ìˆ˜ì •ëœ ì½”ë“œë¥¼ ì¶”ì¶œí•˜ì§€ ëª»í–ˆìŠµë‹ˆë‹¤. ìš°ì¸¡ ì‘ë‹µì„ í™•ì¸í•´ì£¼ì„¸ìš”.")

    """ë§ˆí¬ë‹¤ìš´ í…ìŠ¤íŠ¸ë¥¼ HTMLë¡œ ë³€í™˜"""
    def _format_as_html(self, text: str) -> str:
        """ë§ˆí¬ë‹¤ìš´ í…ìŠ¤íŠ¸ë¥¼ HTMLë¡œ ë³€í™˜"""
        try:
            import markdown
            # ë§ˆí¬ë‹¤ìš´ í™•ì¥ ê¸°ëŠ¥ í¬í•¨ (í…Œì´ë¸”, ì½”ë“œ í•˜ì´ë¼ì´íŠ¸ ë“±)
            html_content = markdown.markdown(
                text,
                extensions=['extra', 'nl2br', 'sane_lists']
            )
            # ê¸°ë³¸ ìŠ¤íƒ€ì¼ ì ìš©
            styled_html = f'''
            <div style="color: #d4d4d4; font-family: 'Noto Sans KR', sans-serif; font-size: 11pt; line-height: 1.6;">
                {html_content}
            </div>
            <style>
                strong {{ color: #4ec9b0; font-weight: bold; }}
                em {{ color: #ce9178; font-style: italic; }}
                code {{ background-color: #1e1e1e; color: #d7ba7d; padding: 2px 4px; border-radius: 3px; font-family: 'Consolas', monospace; }}
                h1, h2, h3 {{ color: #569cd6; }}
                ul, ol {{ padding-left: 20px; }}
                blockquote {{ border-left: 3px solid #569cd6; padding-left: 10px; margin-left: 0; color: #9cdcfe; }}
            </style>
            '''
            return styled_html
        except ImportError:
            # markdown ë¼ì´ë¸ŒëŸ¬ë¦¬ê°€ ì—†ìœ¼ë©´ ê¸°ë³¸ HTML ì²˜ë¦¬
            import html
            escaped = html.escape(text)
            formatted = escaped.replace('\n', '<br>')
            return f'<div style="color: #d4d4d4; font-family: \'Noto Sans KR\', sans-serif; font-size: 11pt; line-height: 1.6;">{formatted}</div>'
